Readme.md
# Unmai Zeithi - 2.0

A simple Ethereum DApp for determining authenticity of information
---

##Testnet Deployments

- Ropsten:

    contract address: 0x3Eb5EA89f7Dc26B23Ec6aD3d5700b1cB1e8Da32D

    etherscan: https://ropsten.etherscan.io/address/0x3eb5ea89f7dc26b23ec6ad3d5700b1cb1e8da32d

- Rinkeby:

    contract address: 0x4C6F76db52fb8146A140aEfc09EcBcD0Ca32B055
    
    etherscan: https://rinkeby.etherscan.io/address/0x4C6F76db52fb8146A140aEfc09EcBcD0Ca32B055


## Deploy commands

Use the following commands to clone the project

```bash

git clone https://gitlab.com/shivjidboss/UnmaiZeithi2.0

cd UnmaiZeithi2.0

sudo docker-compose build

sudo docker-compose up

```

The application should be available at `http://localhost:4200`

**NOTE**
To deploy in other networks modify "docker-compose.yml"

---


## Introduction

---

Establishing the reliability of online information is a daunting but critical current challenge. Fake news detection is defined as the prediction of the chances of a particular news article to be intentionally deceiving.  We propose decentralised platform aiming to authenticate the genuinity of news, making it easier to spot  trusted and useful information, preventing the spreading of misinformation. Spreading genuine knowledge to make life better. An attempt to fight fake news.We intend to use *blockchain* for keeping track of ‘*votes*’ garnered by each article .

---

## Components

---

The project contains 4 parts :

1. A Go-Ethereum server 
2. A node.js  Express server
3. A fastify  node.js Server
4. An angular nginx server

---

## Docker Usage

---

### Prerequisites

This project uses docker-compose and Docker containers. If you do not have these installed please follow the instructions here: <https://docs.docker.com/install/>

---

**NOTE**

Any Linux distributions which support Docker should work.

---

### Building Docker containers

Before starting the project make sure the Docker service is up and running.

To start up the environment, perform the following tasks:

a)Open a terminal window.

b)Change your working directory to the same directory where you saved the Docker Compose file.
 
c)Run the following commands:

```bash
sudo docker-compose up --build
#The `docker-compose.yml` file creates the required containers
```

To stop the validator and destroy the containers, type `^c` in the docker-compose window, wait for it to stop, then type

```bash
sudo docker-compose down
```



Docs/changelog.md
# Changelog
All notable changes to this project since version 1.0 will be documented in this file.

## [2.0.0] - 30-07-2019

- Added an option to "Donate" ether to authors as a token of appreciation
- Added a notification pane to view status of various transactions in the DApp.
- Added check for uniqueness of user address while creating account
- Added check for uniquenss of article (using articlehash) while submitting new articles
- Fixed faulty voting method (voting.sol)

UI/truffle/test/main.js
// contract('Main', function(accounts) {
//   it("should assert true", function(done) {
//     var main = Main.deployed();
//     assert.isTrue(true);
//     done();
//   });
// });


UI/truffle/test/test.js
// const assert = require('assert');
// const ganache = require('ganache-cli');
// const Web3 = require('web3');
// const web3 = new Web3(ganache.provider());
// const json = require('./CompiledContracts/Main.json');

// const Main = artifacts.require("Main");



// let accounts;
// let firstNode;
// const interface = json['abi'];
// const bytecode = json['bytecode'];

// describe('Main', accounts => {

//     before(async () => {
//         accounts = await web3.eth.getAccounts();
//         firstNode = accounts[0];
//         Main = await new web3.eth.Contract(interface)
//             .deploy({ data: bytecode, arguments: ['Owner','owner@uz.com'] })
//             .send({ from: firstNode, gas: '6000000' });
//         await Main.methods.newUser(accounts[1], "TestUsr", "test@uz.com");
//     });

//     it('new article', ()=>{ 
//         Main.deployed()
//             .then(instance => instance.getBalance.call(firstNode))
//             .then(balance => {
//                 assert.notEqual(balance.valueOf(), 0, "account balance is 0")
//             });
//     });


// });



tmplt/app.js
const tmplt = require('./templating');
const preview = require('./preview');
const fastify = require('fastify')({ logger: true });
const url = require('url');
const hash = require('object-hash');

fastify.register(require('fastify-cors'), { 
    // put your options here
    origin:"*"
  });

fastify.get('/url', async (request, reply) => {
    console.log("got request");
    const urlParts = url.parse(request.req.url, true);
    const params = urlParts.query;
    const link = params.link;

    console.log("*************",link, "*********");

    const ct = await tmplt(link);
    console.log(ct);
    const page = `
    <!DOCTYPE html>
    <html>
    <head>${ct.head}</head>
    <body>
    
        <div>${ct.body}</div>
    
    </body>
    </html>`;
    reply.header('Content-Type', 'text/html');
    reply.send(page);
});

fastify.get('/preview', async (request, reply) => {
    const urlParts = url.parse(request.req.url, true);
    const params = urlParts.query;
    const link = params.link;
    const ct = await preview(link);
    const data = {
        image : ct.image,
        title : ct.title,
        author : ct.author,
        publisher : ct.publisher,
        description : ct.description,
        content : ct.text
    }
    const hashedData = hash(data);
    const rply = {
        data : data,
        hash : hashedData
    }
   
    reply.header('Content-Type', 'application/json');
    reply.send(rply);
});

// Run the server!
const start = async () => {
    try {
        await fastify.listen(3000, '0.0.0.0')
        fastify.log.info(`server listening on ${fastify.server.address().port}`)
    } catch (err) {
        fastify.log.error(err)
        process.exit(1)
    }
}
start();


UI/truffle/truffle-config.js
/**
 * Use this file to configure your truffle project. It's seeded with some
 * common settings for different networks and features like migrations,
 * compilation and testing. Uncomment the ones you need or modify
 * them to suit your project as necessary.
 *
 * More information about configuration can be found at:
 *
 * truffleframework.com/docs/advanced/configuration
 *
 * To deploy via Infura you'll need a wallet provider (like truffle-hdwallet-provider)
 * to sign your transactions before they're sent to a remote public node. Infura API
 * keys are available for free at: infura.io/register
 *
 * You'll also need a mnemonic - the twelve word phrase the wallet uses to generate
 * public/private key pairs. If you're publishing your code to GitHub make sure you load this
 * phrase from a file you've .gitignored so it doesn't accidentally become public.
 *
 */

const HDWalletProvider = require('truffle-hdwallet-provider');
const infuraKey = "e1bebdf8e8884fa38c96336d6f370913";

const fs = require('fs');
const mnemonic = fs.readFileSync(".secret").toString().trim();
//const mnemonic = "fruit rude trouble ... carry arrange";

module.exports = {
  /**
   * Networks define how you connect to your ethereum client and let you set the
   * defaults web3 uses to send transactions. If you don't specify one truffle
   * will spin up a development blockchain for you on port 9545 when you
   * run `develop` or `test`. You can ask a truffle command to use a specific
   * network from the command line, e.g
   *
   * $ truffle test --network <network-name>
   */

  networks: {
    // Useful for testing. The `development` name is special - truffle uses it by default
    // if it's defined here and no other network is specified at the command line.
    // You should run a client (like ganache-cli, geth or parity) in a separate terminal
    // tab if you use this network and you must also set the `host`, `port` and `network_id`
    // options below to some value.
    //
    // development: {
    //   host: "127.0.0.1",     // Localhost (default: none)
    //   port: 8545,            // Standard Ethereum port (default: none)
    //   network_id: "4002",       // Any network (default: none)
    //   gas: 62831850
    //  },
    development: {
      host: "geth",
      port: 8545,
      network_id: "4002",
      gas: 62831850
    },

    ropsten: {
      provider: new HDWalletProvider(mnemonic, "https://ropsten.infura.io/v3/" + infuraKey),
      network_id: 3,
      gas: 5500000,
      confirmations: 2,
      timeoutBlocks: 200,
      skipDryRun: true
    },  
    
    rinkeby: {
      provider: new HDWalletProvider(mnemonic, "https://rinkeby.infura.io/v3/" + infuraKey),
      network_id: 4,
      gas: 5500000,
      confirmations: 2,
      timeoutBlocks: 200,
      skipDryRun: true
    }, 

    // Another network with more advanced options...
    // advanced: {
      // port: 8777,             // Custom port
      // network_id: 1342,       // Custom network
      // gas: 8500000,           // Gas sent with each transaction (default: ~6700000)
      // gasPrice: 20000000000,  // 20 gwei (in wei) (default: 100 gwei)
      // from: <address>,        // Account to send txs from (default: accounts[0])
      // websockets: true        // Enable EventEmitter interface for web3 (default: false)
    // },

    // Useful for deploying to a public network.
    // NB: It's important to wrap the provider as a function.
    // ropsten: {
      // provider: () => new HDWalletProvider(mnemonic, `https://ropsten.infura.io/${infuraKey}`),
      // network_id: 3,       // Ropsten's id
      // gas: 5500000,        // Ropsten has a lower block limit than mainnet
      // confirmations: 2,    // # of confs to wait between deployments. (default: 0)
      // timeoutBlocks: 200,  // # of blocks before a deployment times out  (minimum/default: 50)
      // skipDryRun: true     // Skip dry run before migrations? (default: false for public nets )
    // },

    // Useful for private networks
    // private: {
      // provider: () => new HDWalletProvider(mnemonic, `https://network.io`),
      // network_id: 2111,   // This network is yours, in the cloud.
      // production: true    // Treats this network as if it was a public net. (default: false)
    // }
  },

  // Set default mocha options here, use special reporters etc.
  mocha: {
    // timeout: 100000
  },

  // Configure your compilers
  compilers: {
    solc: {
      // version: "0.5.1",    // Fetch exact version from solc-bin (default: truffle's version)
      // docker: true,        // Use "0.5.1" you've installed locally with docker (default: false)
      // settings: {          // See the solidity docs for advice about optimization and evmVersion
      //  optimizer: {
      //    enabled: false,
      //    runs: 200
      //  },
      //  evmVersion: "byzantium"
      // }
    }
  }
}


UI/src/app/drafter/drafter.component.ts
import { Component, OnInit } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Router } from "@angular/router";

import { ArticleService } from '../services/article.service';

import { Article } from '../ngDBModels';
import { TemplatingService } from '../services/templating.service';
import { Preview } from '../model';
import { Web3ServiceService } from '../services/web3-service.service';
import { AuthService } from '../services/auth.service';
import { async } from '@angular/core/testing';

@Component({
  selector: 'app-drafter',
  templateUrl: './drafter.component.html',
  styleUrls: ['./drafter.component.css']
})
export class DrafterComponent implements OnInit {

  link = "http://localhost:3000/";
  prvwPlaceholder = "<div class='draftPrvwPlaceholder' style='font-size: 30px;margin-left: auto;"
    +"margin-right:auto;margin: 0;position: relative;text-align: center;top: 45%;'>Article Preview Appears Here</div>";
  private myTmplt : any = this.prvwPlaceholder;
  private artHash = "";
  disablePublish = true;
  artUniq = 0; 
  private newArticle: Article;
  data;
  message = "";

  constructor(
    private http: HttpClient,
    private router: Router,
    private articleService: ArticleService,
    private template: TemplatingService,
    private web3: Web3ServiceService,
    private usr: AuthService
  ) { }

  ngOnInit() {
  }

  async getPreview(artLink){
    this.artUniq = 1;
    const linkTemp = this.link+"url?link="+artLink;
    this.http.get(this.link+"url?link="+artLink,{responseType: 'text'}).subscribe(async (data: any ) => {
      this.myTmplt = data;
      this.artHash = (await this.template.preview(`${this.link}preview?link=${linkTemp}`)).hash;
      this.chkArtUniq(this.artHash);
    });
  }

  async onSubmit(artLink){
    //e.preventDefault();
    /* stop here if form is invalid
    if (this.loginForm.invalid) {
        return;
    }*/
    this.disablePublish = true;
    this.artUniq = 1;
    const usr = this.usr.getUserDetails();
    const time: string = Date.toString();
    const linkTemp = this.link+"url?link="+artLink; //To do get Article link
    const prvw: Preview = await this.template.preview(`${this.link}preview?link=${linkTemp}`);
    const rtrn =  this.web3.newArticle(prvw.hash, artLink, prvw.data.title, usr.userId, time);
    const artId = rtrn.artid;
    
    this.newArticle = {
      artId: artId,
      hash: prvw.hash,
      title: prvw.data.title,
      img: prvw.data.image,
      link: artLink,
      author: prvw.data.author[0],
      text: prvw.data.description,
      authorId: usr.userId,
      votes: [],
      upVotes: 0,
      downVotes: 0
    };

    const tras = await rtrn.function.on('transactionHash', async(th: string) => {
      console.log(th);
      let c = await this.conf(th);
      if(c){
        this.db();
      }
    });
    console.log(tras);
  }

  db = () => {
    console.log(this.newArticle, " about to enter into Db");
    this.articleService.newArticle(this.newArticle)
      .subscribe(
          data => { // 'TODO : prevent redirect and show error msg if failed???
            this.router.navigate(['/']);
          },
          err => {
            this.message = err.error.msg;
          });
  }

  conf = async (th: string) => {
    let cnfrmd : boolean = false;
    let count = 1;
      while(!cnfrmd){
        console.log(count++);
        cnfrmd = await this.web3.transConf(th);
        await this.web3.delay(1000);
      }
      return (cnfrmd);
  }

  //  [2.0]

  chkArtUniq = async (h: string) =>{
    const usr = await this.usr.getUserDetails();
    let b = await this.web3.chkArticleUniq(h, usr.userId);
    if( b == true ){
      this.artUniq = 2;
      this.disablePublish = false;
    }
    else{
      this.artUniq = 3;
      this.disablePublish = true;
    }
  }

 

}


UI/src/app/viewarticle/viewarticle.component.ts
import { Component, OnInit } from '@angular/core';
import { Router, ActivatedRoute} from "@angular/router";
import { HttpClient } from '@angular/common/http'

import { AuthService } from '../services/auth.service';
import { Web3ServiceService } from '../services/web3-service.service';
import { ArticleService } from '../services/article.service'; 

import { Article } from '../model';
import { User } from '../ngDBModels';

@Component({
  selector: 'app-viewarticle',
  templateUrl: './viewarticle.component.html',
  styleUrls: ['./viewarticle.component.css']
})
export class ViewarticleComponent implements OnInit {


  private artLink = "http://localhost:3000/url?link=";
  private artId : string;
  private usr : User;
  private article = "Loading...";
  private art;
  private articleFromDB;
  private upVotes = 154;  
  private downVotes = 26; 
  private upVoted = "";   
  private downVoted = "";
  constructor(
    private http: HttpClient,
    private route: ActivatedRoute,
    private articleService: ArticleService,
    private authService: AuthService,
    private router:Router,
    private web3 : Web3ServiceService
  ) { 
  }


  async ngOnInit() {
    this.usr = this.authService.getUserDetails();
    this.route.queryParams.subscribe(params => {
        this.artId = params.artId;
        this.getArticle(params.link);
        this.articleService.getArticleFromDB(this.artId).subscribe((art)=>{
          this.articleFromDB = art;
          console.log(this.articleFromDB);
        });
      });
    console.log(this.artId);
    this.art = await this.web3.getArticle(this.artId,this.usr.userId.toString());
  
    setTimeout(() => {
      console.log("this.art: "+JSON.stringify(this.art));
    }, 10000);
    
    const vote =await this.web3.getVote(this.artId,this.usr.userId.toString());
    console.log("Vote: "+vote);
    if(vote == 1){
      this.upVoted = "upvoted";
      this.downVoted = "";
    }
    if(vote == 2){
      this.downVoted = "downvoted";
      this.upVoted = "";
    }
  }

  getArticle(link){
    console.log(this.artLink+link);
    this.http.get(this.artLink+link,{responseType: 'text'}).subscribe((data: any ) => {
      this.article = data;
    });
  }

  async checkVotedOrNot(){
    let vote = await this.web3.getVote(this.artId,this.usr.userId.toString());
    console.log("Vote: "+vote);
    if(vote == 1){
      this.upVoted = "upvoted";
      this.downVoted = "";
    }
    else if(vote == 2){
      this.downVoted = "downvoted";
      this.upVoted = "";
    }
    else{
      this.upVoted = "";
      this.downVoted = "";
    }
  }

  upVote(){
    if(!this.authService.currentUserValue){
      this.openModal();      
    }
    else{
      let opt: number;
      if(this.upVoted == "upvoted"){
        opt = 2;    //already upvoted -> cancel upvote
        this.upVoted = "";
        this.downVoted = "";
      } 
      else if(this.downVoted == "downvoted"){
        opt = 5;   //earlier downvoted -> change to upvote
        this.upVoted = "unfinishedUp";
      } 
      else{
        opt = 1;   //upvote
        this.upVoted = "unfinishedUp";
      } 
      const vote = this.web3.vote(true,this.artId,this.usr.userId.toString(), opt);

      vote.on('transactionHash', async(th: string) => {
        console.log(th);
        let c = await this.conf(th);
        if(c){
          this.db(opt);
        }
      });
      vote.on('receipt', (r: any) =>{
        console.log(r);
      });
    }
  }

  downVote(){
    if(!this.authService.currentUserValue){
      this.openModal(); 
    }
    else{
      let opt: number;
      if(this.downVoted == "downvoted"){
        opt = 4;        //already downvoted -> cancel downvote
        this.upVoted = "";
        this.downVoted = "";
      } 
      else if(this.upVoted == "upvoted"){
        opt = 6;       //earlier upvoted -> change to downvote
        this.downVoted = "unfinishedDown";
      }
      else{
        opt = 3;       //downvote
        this.downVoted = "unfinishedDown";
      }
      const vote = this.web3.vote(false,this.artId,this.usr.userId.toString(), opt);

      vote.on('transactionHash', async(th: string) => {
        console.log(th);
        let c = await this.conf(th);
        if(c){
          this.db(opt);
        }
      });
      vote.on('receipt', (r: any) =>{
        console.log(r);
      });

      
      // this.web3.vote(false,this.artId,this.usr.userId.toString()).then(()=>{
      //   this.articleService.voteArticle({artId: this.artId, updtOpt: opt}).subscribe((art)=>{
      //     this.articleFromDB = art;
      //     this.checkVotedOrNot();
      //     this.art = this.web3.getArticle(this.artId,this.usr.userId.toString());
      //   });
      //   console.log(this.articleFromDB);
      // });
    }
  }

  openModal() {
    document.getElementById('openModalButton').click();
  }

  redirectToLogin(){
    this.router.navigate(['/login']);
  }

  redirectToRegister(){
    this.router.navigate(['/register']);
  }

  conf = async (th: string) => {
    let cnfrmd : boolean = false;
    let count = 1;
      while(!cnfrmd){
        console.log(count++);
        cnfrmd = await this.web3.transConf(th);
        await this.web3.delay(1000);
      }
      return (cnfrmd);
  }

  db = (opt: number) => {
    this.articleService.voteArticle({artId: this.artId, updtOpt: opt}).subscribe((art)=>{
      this.articleFromDB = art;
      this.checkVotedOrNot();
      this.art = this.web3.getArticle(this.artId,this.usr.userId.toString());
    });
    console.log(this.articleFromDB);
  }

}


UI/src/app/services/web3-service.service.ts
import { Injectable } from '@angular/core';
import * as web3 from 'web3';
import {Subject} from 'rxjs';
import { ContractJSON, Acnt, Article } from '../model';

import { ArticleService } from './article.service';
import { Globals } from '../globals';

const Web3 = web3.default;

const MyContractJSON: ContractJSON = require('../../../truffle/build/contracts/Main.json');

// const contractAddress = MyContractJSON.networks['4002'].address;
const contractAddress = MyContractJSON.networks[4002].address;
const abi = MyContractJSON.abi;
const web3Provider =new Web3.providers.HttpProvider("http://127.0.0.1:8545");

declare let window: any;

@Injectable({
  providedIn: 'root'
})
export class Web3ServiceService {
  
  private myWeb3 :any;
  public UZ :any;
  public myAccounts: string[] = [];
  public myAccountsObservable = new Subject<String[]>();
  
  constructor(private articleService: ArticleService, private globals: Globals) {
    this.startWeb3();
   }

  startWeb3 = () => {
    console.info("Starting Web3...");
    console.log("Contract Address ->",contractAddress);
    this.myWeb3 = new Web3(web3Provider);
    this.UZ = new this.myWeb3.eth.Contract(abi, contractAddress);
    if(!this.myWeb3.eth.net.isListening()){
      console.warn("Couldn't connect to any web3 provider!!!")
      return;
    }
    setTimeout(() => this.getAccounts(), 200); // changed to timeout as it was refreshing every 0.2 sec
  };

   getAccounts = () => {
    this.myWeb3.eth.getAccounts((error,accounts) =>{
      if(error){
        console.warn('There was an error fetching your accounts.');
        return;
      }
      console.log('getting accounts');
      // console.log(accounts);

      if(accounts.length === 0 ){
        console.warn("Couldn't get any accounts!!!Error!!!");
      }
      else if( (this.myAccounts.length !== 0 || this.myAccounts.length !== accounts.length || this.myAccounts[0] !== accounts[0] ) ){
        this.myAccountsObservable.next(accounts);
        this.myAccounts = accounts;
      }
    });
  }

  getArticle = async (artAddress: string, userAddress: string):Promise<Article> => {
    
    let article : Article = await this.UZ.methods.getArticle(artAddress).call({from: userAddress}).then((v: Article) => {
      return v;
    });
    return article;
  }

  getUser = async (userAddress: string) => {
    let u = await this.UZ.methods.getUser(userAddress).call({from: userAddress}).then((v) => {
      return v;
    });
    return u;
  }


  // vote = async (vote: Boolean, artAddress: string, userAddress: string) => {
  //   let cnfrmd: boolean = false;
  //   await this.UZ.methods.voteArticle(vote, artAddress).send({from: userAddress.toString(), gas: 6000000})
  //   .on('transactionHash', async (tHash: string) => {
  //     console.log(tHash);
  //     let count = 1;
  //     while(!cnfrmd){
  //       cnfrmd =await this.transConf(tHash);
  //       await this.delay(1000);
  //       console.log(count++);
  //     }
  //   });
  //   return cnfrmd;
  // }

  getVote = async (artAddress: string, userAddress: String) => {
    const myVote: number = await this.UZ.methods.getVote(artAddress).call({from: userAddress});
    return myVote;
  }

  transConf = async (tranHash : string) : Promise<boolean> => {
    try{
      const tresp = await this.myWeb3.eth.getTransaction(tranHash);
      return tresp.blockNumber === null ? false : true;
    }
    catch(e){
      console.log(e);
      return false;
    }
  }

  delay = async (ms: number) => {
    return new Promise((resolve) => {
      setTimeout(resolve, ms);
    });
  }

  // d = async (f: boolean) => {
  //   return new Promise(async (resolve) => {
  //     while(!f){
  //       f = await this.transConf(tHash);
  //     }
  //     resolve;      
    // });
  // }



  // [2.0]


  donate = async (userAddress: string, toName: String, toAddr: string, amnt:string) => {
    this.UZ.methods.donate(toAddr).send({from: userAddress, to: toAddr, gas: 6000000, value: this.myWeb3.utils.toWei(amnt.toString(), "ether")})
      .on('transactionHash', async (tHash: string) => {
        this.globals.Notifies.push(
          {
            time: String(Date.now()), 
            status: "Tx. hash generated", 
            statCode: 1, 
            msg: "Donation to "+toName+" of "+amnt+" ether", 
            txHash: tHash, 
            msgHist: [{}]
          });
        console.log(tHash);
      })
      .on('receipt', async (tranReceipt) =>{
        let t = this.globals.Notifies.findIndex(i=>i.txHash == tranReceipt.transactionHash);
        this.globals.Notifies[t].msgHist.push({oldm: this.globals.Notifies[t].status, t: this.globals.Notifies[t].time, stat: this.globals.Notifies[t].statCode});
        this.globals.Notifies[t].time = String(Date.now());
        this.globals.Notifies[t].statCode = 2;
        this.globals.Notifies[t].status = "Tx. receipt generated";
        console.log(tranReceipt);
        return true;
      })
      .on('error', (err)=>{

        let t = this.globals.Notifies.findIndex(i=>i.txHash == err.transactionHash);
        if(t){
          this.globals.Notifies[t].msgHist.push({oldm: this.globals.Notifies[t].status, t: this.globals.Notifies[t].time, stat: this.globals.Notifies[t].statCode});
          this.globals.Notifies[t].time = String(Date.now());
          this.globals.Notifies[t].statCode = 3;
          this.globals.Notifies[t].status = "Tx. failed";
          console.log(err);
          return false;
        }
        else{
          this.globals.Notifies.push(
            {
              time: String(Date.now()), 
              status: "Tx. failed", 
              statCode: 3, 
              msg: "Donation to "+toName+" of "+amnt+" ether", 
              txHash: "---", 
              msgHist: [{}]
            });
          console.log(err);
          return false;
        }
      });
    }

    // .then((tx) => {
    //   console.log(tx);
    //   return true;
    // }).catch((err)=>{
    //   console.log(err);
    //   return false;
    // });


  vote =  (vote: Boolean, artAddress: string, userAddress: string, voteUD: Number) => {
    return this.UZ.methods.voteArticle(vote, artAddress).send({from: userAddress.toString(), gas: 6000000})
    .on('transactionHash', async (tHash: string) => {
      let vStat;
      if(voteUD == 1 || voteUD == 5)vStat="Up";
      else if(voteUD == 3 || voteUD == 6)vStat="Down";
      else vStat = "Cancel ";
      this.globals.Notifies.push(
        {
          time: String(Date.now()), 
          status: "Tx. hash generated", 
          statCode: 1, 
          msg: vStat+"vote on article", 
          txHash: tHash, 
          msgHist: [{}]
        });
      console.log(tHash);
    })
    .on('receipt', async (tranReceipt) =>{
      let t = this.globals.Notifies.findIndex(i=>i.txHash == tranReceipt.transactionHash);
      this.globals.Notifies[t].msgHist.push({oldm: this.globals.Notifies[t].status, t: this.globals.Notifies[t].time, stat: this.globals.Notifies[t].statCode});
      this.globals.Notifies[t].time = String(Date.now());
      this.globals.Notifies[t].statCode = 2;
      this.globals.Notifies[t].status = "Tx. receipt generated";
      console.log(tranReceipt);
      return true;
    })
    .on('error', (err)=>{

      let t = this.globals.Notifies.findIndex(i=>i.txHash == err.transactionHash);
      if(t>-1){
        this.globals.Notifies[t].msgHist.push({oldm: this.globals.Notifies[t].status, t: this.globals.Notifies[t].time, stat: this.globals.Notifies[t].statCode});
        this.globals.Notifies[t].time = String(Date.now());
        this.globals.Notifies[t].statCode = 3;
        this.globals.Notifies[t].status = "Tx. failed";
        console.log(err);
        return false;
      }
      else{
        this.globals.Notifies.push(
          {
            time: String(Date.now()), 
            status: "Tx. failed", 
            statCode: 3, 
            msg: (vote==true)?"Up":"Down"+"Vote article", 
            txHash: "---", 
            msgHist: [{}]
          });
        console.log(err);
        return false;
      }
    });
  }

  newArticle =  ( artHash: string, link: string, title: string, userAddress: string, time: string) => {
    let article: Acnt = this.myWeb3.eth.accounts.create(this.myWeb3.utils.randomHex(32));
    let func = this.UZ.methods.newArticle(article.address, artHash, link, userAddress, userAddress, time)
      .send({from: userAddress, gas: 6000000})
      .on('transactionHash', async (tHash: string) => {
        this.globals.Notifies.push(
          {
            time: String(Date.now()), 
            status: "Tx. hash generated", 
            statCode: 1, 
            msg: "New article: "+title, 
            txHash: tHash, 
            msgHist: [{}]
          });
        console.log(tHash);
      })
      .on('receipt', async (tranReceipt) =>{
        let t = this.globals.Notifies.findIndex(i=>i.txHash == tranReceipt.transactionHash);
        this.globals.Notifies[t].msgHist.push({oldm: this.globals.Notifies[t].status, t: this.globals.Notifies[t].time, stat: this.globals.Notifies[t].statCode});
        this.globals.Notifies[t].time = String(Date.now());
        this.globals.Notifies[t].statCode = 2;
        this.globals.Notifies[t].status = "Tx. receipt generated";
        console.log(tranReceipt);
        return true;
      })
      .on('error', (err)=>{

        let t = this.globals.Notifies.findIndex(i=>i.txHash == err.transactionHash);
        if(t>-1){
          this.globals.Notifies[t].msgHist.push({oldm: this.globals.Notifies[t].status, t: this.globals.Notifies[t].time, stat: this.globals.Notifies[t].statCode});
          this.globals.Notifies[t].time = String(Date.now());
          this.globals.Notifies[t].statCode = 3;
          this.globals.Notifies[t].status = "Tx. failed";
          console.log(err);
          return false;
        }
        else{
          this.globals.Notifies.push(
            {
              time: String(Date.now()), 
              status: "Tx. failed", 
              statCode: 3, 
              msg: "New article:"+title, 
              txHash: "---", 
              msgHist: [{}]
            });
          console.log(err);
          return false;
        }
      });
    return {
      artid: article.address,
      function: func
    }
  }

  
  newUser =  (userAddress: string, name: string, emai: string) => {
    return this.UZ.methods.newUser(userAddress, name, emai).send({from: userAddress, gas: 6000000})
    .on('transactionHash', async (tHash: string) => {
      this.globals.Notifies.push(
        {
          time: String(Date.now()), 
          status: "Tx. hash generated", 
          statCode: 1, 
          msg: "New user: "+name, 
          txHash: tHash, 
          msgHist: [{}]
        });
      console.log(tHash);
    })
    .on('receipt', async (tranReceipt) =>{
      let t = this.globals.Notifies.findIndex(i=>i.txHash == tranReceipt.transactionHash);
      this.globals.Notifies[t].msgHist.push({oldm: this.globals.Notifies[t].status, t: this.globals.Notifies[t].time, stat: this.globals.Notifies[t].statCode});
      this.globals.Notifies[t].time = String(Date.now());
      this.globals.Notifies[t].statCode = 2;
      this.globals.Notifies[t].status = "Tx. receipt generated";
      console.log(tranReceipt);
      return true;
    })
    .on('error', (err)=>{
      let t = this.globals.Notifies.findIndex(i=>i.txHash == err.transactionHash);
      if(t>-1){
        this.globals.Notifies[t].msgHist.push({oldm: this.globals.Notifies[t].status, t: this.globals.Notifies[t].time, stat: this.globals.Notifies[t].statCode});
        this.globals.Notifies[t].time = String(Date.now());
        this.globals.Notifies[t].statCode = 3;
        this.globals.Notifies[t].status = "Tx. failed";
        console.log(err);
        return false;
      }
      else{
        this.globals.Notifies.push(
          {
            time: String(Date.now()), 
            status: "Something went wrong. Tx. failed", 
            statCode: 3, 
            msg: "New user: "+name, 
            txHash: "---", 
            msgHist: [{}]
          });
        console.log(err);
        return false;
      }
    })
    // .then((tranReceipt) =>{
    //   this.globals.Notifies.push(
    //     {
    //       time: String(Date.now()), 
    //       status: name+" is now part of UZ!", 
    //       statCode: 1, 
    //       msg: "New user: "+name, 
    //       txHash: tranReceipt, 
    //       msgHist: [{}]
    //     });
    // })
    // .catch((e)=>{
    //   console.log(e);
    //   this.globals.Notifies.push(
    //     {
    //       time: String(Date.now()), 
    //       status: "Something went wrong. Tx. failed", 
    //       statCode: 3, 
    //       msg: "New user: "+name, 
    //       txHash: "---", 
    //       msgHist: [{}]
    //     });
    // });
  }

  chkArticleUniq = async (h: string, userAddress: string) => {
    let u = this.UZ.methods.chkArticleUniq(h).call({from: userAddress}).then((b) =>{
      return  b;
    });
    return u;
  }
  

}



// tranReceipt = async (tranHash: String) : Promise<boolean> => {
  //   try{
  //     const tresp = await this.myWeb3.eth.getTransactionReceipt(tranHash, (e, data) => {
  //       if (e !== null) {
  //           console.log("Could not find a transaction for your id! ID you provided was " + tranHash);
  //       } 
  //       else {
  //           console.log(data);
  //           if(data.status == '0x0') {
  //               console.log("The contract execution was not successful, check your transaction !");
  //               return false;
  //           } else {
  //               console.log("Execution worked fine!");
  //               return true;
  //           }
  //       }
  //     });
  //   }
  //   catch(e){
  //     console.log(e);
  //     return false;
  //   }
  // }